bgpd/bgp_mplsvpn.c

```diff
@@ -1336,6 +1336,230 @@ leak_update(struct bgp *to_bgp, struct bgp_dest *bn,
 	return new;
 }
 
+void bgp_mplsvpn_path_nh_label_unlink(struct bgp_path_info *pi)
+{
+	struct bgp_label_per_nexthop_cache *blnc;
+
+	if (!pi)
+		return;
+
+	blnc = pi->label_nexthop_cache;
+
+	if (!blnc)
+		return;
+
+	LIST_REMOVE(pi, label_nh_thread);
+	pi->label_nexthop_cache->path_count--;
+	pi->label_nexthop_cache = NULL;
+
+	if (LIST_EMPTY(&(blnc->paths)))
+		bgp_label_per_nexthop_free(blnc);
+}
+
+/* Called upon reception of a ZAPI Message from zebra, about
+ * a new available label.
+ */
+static int bgp_mplsvpn_get_label_per_nexthop_cb(mpls_label_t label,
+						void *context, bool allocated)
+{
+	struct bgp_label_per_nexthop_cache *blnc = context;
+	mpls_label_t old_label;
+	int debug = BGP_DEBUG(vpn, VPN_LEAK_LABEL);
+	struct bgp_path_info *pi;
+	struct bgp_table *table;
+
+	old_label = blnc->label;
+
+	if (debug)
+		zlog_debug("%s: label=%u, allocated=%d, nexthop=%pFX", __func__,
+			   label, allocated, &blnc->nexthop);
+	if (allocated)
+		/* update the entry with the new label */
+		blnc->label = label;
+	else
+		/*
+		 * previously-allocated label is now invalid
+		 * eg: zebra deallocated the labels and notifies it
+		 */
+		blnc->label = MPLS_INVALID_LABEL;
+
+	if (old_label == blnc->label)
+		return 0; /* no change */
+
+	/* update paths */
+	if (blnc->label != MPLS_INVALID_LABEL)
+		bgp_zebra_send_nexthop_label(ZEBRA_MPLS_LABELS_ADD, blnc->label,
+					     ZEBRA_LSP_BGP, &blnc->nexthop);
+
+	LIST_FOREACH (pi, &(blnc->paths), label_nh_thread) {
+		if (!pi->net)
+			continue;
+		table = bgp_dest_table(pi->net);
+		if (!table)
+			continue;
+		vpn_leak_from_vrf_update(blnc->to_bgp, table->bgp, pi);
+	}
+
+	return 0;
+}
+
+/* get label for a given bgp update
+ * return the label value:
+ * - per-vrf value:
+ *      if nexthop of path is not valid or there are no nexthops
+ *      if nexthop of path is not connected
+ *      if path becomes ECMP
+ *      if blackhole is found
+ * - valid per-nexthop value
+ *      if nexthop of path is valid and connected and a label has been obtained
+ * - invalid value
+ *      if nexthop of path is valid and connected and label has not yet been
+ * obtained
+ */
+static mpls_label_t bgp_mplsvpn_get_path_info_label(struct bgp_path_info *pi,
+						    struct bgp *to_bgp,
+						    struct bgp *from_bgp,
+						    afi_t afi, safi_t safi)
+{
+	struct bgp_nexthop_cache *bnc;
+	struct bgp_label_per_nexthop_cache *blnc;
+	struct bgp_label_per_nexthop_cache_head *tree;
+	struct prefix *nh_pfx = NULL;
+	struct prefix nh_resolved = {0};
+	struct bgp *bgp_nexthop = NULL;
+	afi_t nh_afi;
+
+	/* check the next-hop reachability in the
+	 * bgp instance where the next-hop of the entry is
+	 */
+	if (pi->extra && pi->extra->bgp_orig)
+		bgp_nexthop = pi->extra->bgp_orig;
+	else
+		bgp_nexthop = from_bgp;
+
+	nh_afi = BGP_ATTR_NH_AFI(afi, pi->attr);
+
+	if (bgp_find_or_add_nexthop(from_bgp, bgp_nexthop, nh_afi, safi, pi,
```

Comment:
what if `pi->label_nexthop_cache` is set. Do we need to unlink ?

Reply:
bgp_find_or_add_nexthop() is later called in leak_update_nexthop_valid() that is called by vpn_leak_from_vrf_update() then leak_update()

Do you really need to test that here ?

Reply:
We need to call the ```bgp_find_or_add_nexthop()``` before doing ```leak_update()```, because we need to know if there is a reachable next-hop associated to that path_info. As we need a valid label, before doing ```leak_update()```, this step is necessary.

Reply:
> what if `pi->label_nexthop_cache` is set. Do we need to unlink ?

actually, there are cases where a reachable BGP update could become unreachable (eg: the nexthop attribute of a prefix becomes not reachable), so the answer is yes, we need to unlink.